[docs] Derived objects (#23462)
MystenLabs/sui: docs/content/concepts/sui-move-concepts/derived-objects.mdx

---
title: Derived Objects
description: Derived objects enable you to deterministically compute object IDs from a parent object ID plus a key, creating a per-parent namespace.
keywords: [ Derived objects, registry, ]
---

Derived objects enable you to deterministically compute object IDs from a parent object ID plus a key, creating a per-parent namespace. This gives you the benefits of a registry (unique slots per key) without storing child objects inside the parent.

Classic registries (dynamic fields) require a parent registry object that owns children. Because mutation of a child typically requires a mutable borrow through the parent, unrelated writes get funneled through the parent. This can create a contention/consensus hotspot if the parent is shared.

> Derived objects keep the namespace (parent+key) but the derived object is a top-level object (not stored in the parent). It can be owned (fast path), shared, or immutable. You avoid sequencing unrelated mutations through the parent, and still guarantee uniqueness per (parent, key).

Reviewer: ```suggestion
Derived objects keep the namespace (parent+key) but the derived object is a top-level object (not stored in the parent). It can be owned (fast path), shared, or immutable. You avoid sequencing unrelated mutations through the parent (for all operations besides creation), while maintaining the guaranteed uniqueness per (parent, key).
```

**On-chain benefits**

-	Less contention and better parallelism: No parent bottleneck for unrelated keys.
-	Cheaper when owned: You can make each derived object owned, avoiding shared-object consensus for most operations.
-	Deterministic uniqueness: One object per (parent, key) without extra bookkeeping.
-	Top-level ergonomics: Clean capability patterns, Object Display, and simpler permissioning.

**Off-chain benefits**

-	Fewer hops: Clients can compute or look up the derived ID directly instead of going through the parent.

> -	Better discoverability/indexing: Explorers/SDKs can infer or cache by (parent, key); fewer sequential queries.

Author: I see what you mean. But I think they point it out from different angles. And three things is always better than one thing.

-	Lean SDK calls: Compute ID and fetch object, instead of mutably touching a parent.

## Derived objects and dynamic fields matrix

The following matrix highlights some of the differences between using dynamic fields versus derived objects. Considering the tradeoffs helps you select the optimum approach for your project.

| Aspect | Derived objects | Dynamic fields |
| --- | --- | --- |
| Address predictability | ✅ Yes | ✅ Yes |
| Parent required? | Only to create | ✅ Yes |
| Ownership type | Any. Can be wrapped or shared, owned, party, or frozen. | Cannot be independently owned. Owner is always the parent. |
| Supports receiving objects? | ✅ Yes | ❌ No |
| Object parallelism | ✅ Yes | Limited. All writes sequenced through parent. |

> | Cost overhead | Double the cost of dynamic fields. Requires UID and dynamic field (to prevent duplicate claims) creation. Receive a rebate upon UID deletion, but no rebate on the dynamic field. | Cheaper and offers full storage rebates on deletion. |

Reviewer: ```suggestion
| Cost overhead | Double the cost of dynamic fields. Requires UID and dynamic field (to prevent duplicate claims) creation. Receive a rebate upon UID deletion, but no rebate on the dynamic field. | Cheaper and offers full storage rebates on deletion. |
```

I think we can remove the cost overhead altogether. My example was also wrong... for UID + UID (for TTO) case, the cost is the same...

> | Loading type | Static. Direct access after creation. | Dynamic. Transactions happen through the parent. |

Reviewer: ```suggestion
| Loading type | Static. Direct access after creation. | Dynamic. They are loaded through the parent. |
```

keywords: [ Derived objects, deterministic addressing, Transfer-to-Object, TTO, parallelization, object composition, scalability ]
---

Derived objects enable you to deterministically compute object IDs from a parent object ID plus a key, unlocking powerful patterns for building scalable, composable systems on Sui. The core insight is that **object addresses can be known before creation**, enabling pre-computation of destinations, guaranteed uniqueness per key, and seamless parallelization of operations.

## Core capabilities

Derived objects provide four fundamental capabilities:

### Deterministic addressing

Object IDs can be computed ahead of time using `derived_object::derive_address(parent_id, key)`. This means applications can predict where objects will live before they exist, enabling sophisticated coordination patterns and reducing the need for on-chain lookups.

### Transfer-to-Object (TTO) compatibility

Because addresses are computable, objects can receive transfers even before they exist. Send assets to a derived address, then create the object later to claim them. This enables patterns like:

- Pre-funding accounts before user onboarding.

> - Conditional object creation based on received assets.

Reviewer: Not sure I follow this one? What did you have in mind?

- Cross-chain bridging to deterministic destinations.

### One-per-key uniqueness

Each `(parent, key)` pair maps to exactly one object address. This gives you registry-like uniqueness guarantees without the registry bottleneck. Perfect for soulbound tokens, per-user configurations, or any scenario requiring unique slots.

### Native parallelization

Unlike dynamic fields that funnel operations through a parent object, derived objects operate independently after creation. Unrelated keys can be modified in parallel, eliminating consensus hotspots while maintaining namespace guarantees.

These capabilities combine to enable entirely new design patterns that were previously impossible or inefficient on other blockchains.

**On-chain benefits**

-	Less contention and better parallelism: No parent bottleneck for unrelated keys.
-	Deterministic uniqueness: One object per (parent, key) without manual bookkeeping.
-	Top-level object ergonomics: Clean capability patterns, Object Display, and simpler permissioning.

**Off-chain benefits**

> -	Fewer hops: Clients can compute or look up the derived ID directly instead of going through the parent.

Reviewer: ```suggestion
-	Fewer hops: Clients can compute the derived object ID directly without requiring a sequential operation to find it through the parent.
```
